contracts/BondToken.sol

```diff
@@ -0,0 +1,80 @@
+
+import "./HubPoolInterface.sol";
+import "./WETH9.sol";
+
+interface _HubPool is HubPoolInterface {
+    // Specify the automatically-implemented rootBundleProposer() getter.
+    function rootBundleProposal() external pure returns (HubPoolInterface.RootBundle memory);
+}
+
+/**
+ * @notice Across Bond Token (ABT).
+ * ABT is a simple deposit contract based on WETH9. ABT is issued proportionally to any address that deposits Ether. It
+ * imposes address-based permissioning on the WETH9 transferFrom() function in order to constrain the movement of ABT
+ * into the Across v2 HubPool contract. When configured as the required HubPool bond token, ABT can dramatically reduce
+ * the attack surface of the HubPool by requiring that addresses are explicitly approved before they can successfully
+ * submit a root bundle proposal. The address-based permissiong does not constrain transfers that are needed to dispute
+ * a root bundle proposal, so the ability of decentralised/unknown actors to dispute is unaffected.
+ */
+contract BondToken is WETH9, Ownable {
+    using Address for address;
+
+    _HubPool public immutable hubPool;
+
+    /**
+     * @notice Addresses that are permitted to make HubPool root bundle proposals.
+     */
+    mapping(address => bool) public proposers;
+
+    /**
+     * @notice Emitted on proposer permissions update.
+     */
+    event ProposerModified(address proposer, bool enabled);
+
+    /**
+     * @notice BondToken constructor.
+     * @param _hubPool Address of the target HubPool contract.
+     */
+    constructor(_HubPool _hubPool) {
+        name = "Across v2 Bond Token";
+        symbol = "ABT";
+        hubPool = _hubPool;
+    }
+
+    /**
+     * @notice Enable or disable an address as an allowed proposer. Emits a "ProposerModified" event on completion.
+     * @param proposer Proposer address to modify.
+     * @param enabled Boolean controlling whether the address is permitted to propose.
+     */
+    function setProposer(address proposer, bool enabled) external onlyOwner {
+        proposers[proposer] = enabled;
+        emit ProposerModified(proposer, enabled);
+    }
+
+    /**
+     * @notice Transfer amt from src to dst. Prevents unauthorised root bundle proposals by blocking transfers to the
+     * HubPool under the following conditions:
+     * - The src address is not a pre-approved proposer, *and*
+     * - The src address is the current proposer of a HubPool root bundle.
+     * Falls back to the base implementation after verifying that the transfer is permitted.
+     * @param src Source address.
+     * @param dst Destination address.
+     * @param amt Amount to transfer.
+     * @return True on success.
+     */
+    function transferFrom(
+        address src,
+        address dst,
+        uint256 amt
+    ) public override returns (bool) {
+        if (dst == address(hubPool)) {
+            require(proposers[src] || hubPool.rootBundleProposal().proposer != src, "Transfer not permitted");
```

Comment:
Does this mean proposer's can't dispute themselves? I don't think we should block such behavior

Reply:
This check only requires that one of the conditions is true, so proposers can always dispute without impediment. There's a test that verifies that - see test/BondToken.e2e.ts line 72.

Reply:
This is _extremely_ counterintuitive, but it's really clever. This is the sort of logic that I think Certora would be really good for verifying. We would probably want to verify that any address with tokens can _always_ send a dispute.

I'm not super worried about this edge case, but this logic does seem to mean that if I proposed and then I was removed from the whitelist, I would then be no longer able to dispute myself.

Just out of curiousity, why did you not use the the check we use in the HubPool, that [unclaimedPoolRebalanceLeafCount](https://github.com/across-protocol/contracts-v2/blob/598b17f019f80540e7d5d0ae7bad15235edb1e62/contracts/HubPoolInterface.sol#L68) != 0 to detect disputes vs proposals. See the [noActiveRequests](https://github.com/across-protocol/contracts-v2/blob/598b17f019f80540e7d5d0ae7bad15235edb1e62/contracts/HubPool.sol#L173-L176) modifier.

EDIT: That last idea doesn't work. I didn't quite get the nuance you picked up on, that the variables are written _before_ the transfer. I think the way you've done it and checking the timestamp is probably the only way I could see this working. Very clever.

Reply:
> I'm not super worried about this edge case, but this logic does seem to mean that if I proposed and then I was removed from the whitelist, I would then be no longer able to dispute myself.

That was also my intuition, but then I tried to implement a test for it and found the opposite - it works unexpected. This is similarly because the pending root bundle is deleted before the bond token transferFrom() function is called. It's a happy accident, I guess.

Reply:
> > I'm not super worried about this edge case, but this logic does seem to mean that if I proposed and then I was removed from the whitelist, I would then be no longer able to dispute myself.
> 
> That was also my intuition, but then I tried to implement a test for it and found the opposite - it works unexpectedly. This is similarly because the pending root bundle is deleted before the bond token transferFrom() function is called. It's a happy accident, I guess.
> 
> For reference, see the test at line 78 in test/BondToken.e2e.ts.

> Just out of curiousity, why did you not use the the check we use in the HubPool, that [unclaimedPoolRebalanceLeafCount](https://github.com/across-protocol/contracts-v2/blob/598b17f019f80540e7d5d0ae7bad15235edb1e62/contracts/HubPoolInterface.sol#L68) != 0 to detect disputes vs proposals. See the [noActiveRequests](https://github.com/across-protocol/contracts-v2/blob/598b17f019f80540e7d5d0ae7bad15235edb1e62/contracts/HubPool.sol#L173-L176) modifier.

That was my first attempt as well, but it doesn't work. At the time of evaluation, the proposed root bundle _is_ the version submitted by the proposer, so it seems reliably non-zero in case of both proposals and disputes.

Reply:
> This is _extremely_ counterintuitive

I've added a @dev comment to the BondToken.safeTransferFrom() implementation to call this out more explicitly. See https://github.com/across-protocol/contracts-v2/pull/226/commits/5215b60ab414e4d1d0f9feac6ab4a0b3aba4fea6.

Reply:
I had the same question as Matt but see your point now!

Reply:
Wow, it keeps getting weirder @pxrl. Really would love some validation on this!